discovery-provider/src/queries/get_health.py

```diff
@@ -469,50 +452,84 @@ def get_latest_chain_block_set_if_nx(redis=None, web3=None):
 
     :rtype (int, string)
     """
+    latest_block_num, latest_block_hash = get_latest_chain_block_from_redis(redis)
 
-    latest_block_num = None
-    latest_block_hash = None
+    if latest_block_num is None or latest_block_hash is None:
+        latest_block_num, latest_block_hash = get_latest_chain_block_from_web3(web3)
+        set_latest_chain_block_in_redis(redis, latest_block_num, latest_block_hash)
 
-    if redis is None or web3 is None:
-        raise Exception("Invalid arguments for get_latest_chain_block_set_if_nx")
+    return latest_block_num, latest_block_hash
 
-    # also check for 'eth' attribute in web3 which means it's initialized and connected to a provider
-    if not hasattr(web3, "eth"):
-        raise Exception(
-            "Invalid web3 argument for get_latest_chain_block_set_if_nx, web3 is not initialized"
-        )
+def get_latest_chain_block_from_web3(web3):
+    latest_block_num = None
+    latest_block_hash = None
 
-    stored_latest_block_num = redis.get(latest_block_redis_key)
-    if stored_latest_block_num is not None:
-        latest_block_num = int(stored_latest_block_num)
+    try:
+        if web3 is None:
+            raise Exception("Web3 instance not valid")
 
-    stored_latest_blockhash = redis.get(latest_block_hash_redis_key)
-    if stored_latest_blockhash is not None:
-        latest_block_hash = stored_latest_blockhash.decode("utf-8")
+        # Also check for 'eth' attribute in web3 which means it's initialized and connected to a provider
+        if not hasattr(web3, "eth"):
+            raise Exception("Invalid web3 argument. web3 is not initialized")
```

Comment:
i think we actually dont want to raise exceptions in this flow. think about the scenario we have a POA gateway outage but we want discovery nodes to keep serving requests. in this case if it throws an exception none of the nodes will be healthy and nothing will get selected (maybe as a fallback something might get selected). but i think the intent here is to not block health check from returning a success due to an outage from a dependency like POA or redis. it should return whatever value it can from the db fallback

Reply:
I believe the code here does what you described. the raised exceptions in line 469 and 473 will be caught in the except block and a warning will be printed. this is also the case for redis. so, at the very end, if redis and web3 fail, we will see `None` for the final block data

Reply:
sorry, its late on a friday evening. you are absolutely right